### PR TITLE
CachingPool: free browser when reset failed

### PR DESCRIPTION
**Pull request by j0tunn**

### CODE REVIEW

**Comment by j0tunn** [comment]
@SwinX @sipayRT @SevInf

**Comment by SwinX** [comment]
:ok:

File: lib/browser-pool/caching-pool.js [hunk 1]
```diff
@@ -35,7 +36,13 @@ CachingPool.prototype.getBrowser = function(id) {
     }
 
     log('has cached browser %o', browser);
-    return browser.reset().thenResolve(browser);
+    return browser.reset()
+        .fail(function(e) {
+            var reject = q.reject.bind(null, e);
+            return this.freeBrowser(browser)
```
Comment (SevInf) [inline lib/browser-pool/caching-pool.js]: `this.freeBrowser` will put it into the cache which is probably not what we want here. Better to call `underlyingPool.freeBrowser` here

Comment (j0tunn) [inline lib/browser-pool/caching-pool.js]: we've got it from the cache, so we put it back. There can be some temporary problems with network or else.

Comment (j0tunn) [inline lib/browser-pool/caching-pool.js]: But probably you are right. Let this be the preflight test: if we can't reset browser session it probably broken, so discard this session and open new one.

**Comment by j0tunn** [comment]
:up:

**Pull request merged**